### PR TITLE
roachprod: only wait on default target cluster if setting backup schedule

### PR DESCRIPTION
**Pull request by ajstorm**
roachprod only uses the default target cluster setting when setting a backup schedule. As a result, if we're not setting a backup schedule in a given start call, we shouldn't wait until the default target cluster setting is set.

Release note: None

Informs: #117926

### CODE REVIEW

**Comment by renatolabs** [comment]
I'd like to understand the issue a little better. I was trying to reproduce it by recreating the situation described in #117926, but couldn't.

> in cases where the cluster has already been setup properly, and is just restarting, AND the user has set `server.controller.default_target_cluster` to something other than `application`, startup will fail

I created a local cluster and I changed the default target cluster to something else:

```
root@localhost:29000/system/defaultdb> SHOW CLUSTER SETTING server.controller.default_target_cluster;                                                                                                           
  server.controller.default_target_cluster
--------------------------------------------
  my-tenant
(1 row)
```

Then stopped my cluster:

```console
% roachprod stop local --sig 15 --wait
local: stopping and waiting 1/1
```

And tried restarting it:

```
% ./bin/roachprod start local --secure --args --config-profile=replication-source
15:11:11 cluster_synced.go:1805: local: checking /Users/rc/local/1/certs.tar
15:11:11 cockroach.go:386: local: starting nodes
15:11:12 cockroach.go:1004: waiting for default target cluster
local: executing sql 1/1
local: executing sql 1/1
local: executing sql 1/1
local: executing sql 1/1
local: executing sql 1/1
15:11:50 cockroach.go:1103: log into DB console with user=roach password=system
15:11:50 cockroach.go:1109: local: setting cluster settings
```

which works.

I was also confused by this comment in the issue:

> roachprod is waiting for  `server.controller.default_target_cluster` to be set to `application`, which never happens.  

Roachprod is not waiting for an `application` tenant; it's just waiting for the default virtual cluster to not be `system`:

https://github.com/cockroachdb/cockroach/blob/b79dcef4589c3e5af2ec829e5d2c56fee9bbe140/pkg/roachprod/install/cockroach.go#L1020-L1022

Which brings us to, I believe, the root of the issue. `cct-232`'s default virtual cluster _is_ currently set to `system`:

```console
$ roachprod sql cct-232:1 --secure --cluster system -- -e "SHOW CLUSTER SETTING server.controller.default_target_cluster;"
  server.controller.default_target_cluster
--------------------------------------------
  system
(1 row)
```

*****

TL;DR: IMO, starting cockroach with `--config-profile=replication-source` **and** manually setting the cluster's default virtual cluster to be `system` is incompatible (since the command line flag implies a deployment model where `system` is not the default cluster).

Is it possible to change the restart script to not pass the config-profile flag?

**Comment by ajstorm** [comment]
Thanks for the review @renatolabs.  Yes, you're right, and I was imprecise in my drafting of the original issue.  You're correct that we're waiting for `server.controller.default_target_cluster` to not be set to `system` (in the common case however, this will mean that it's set to `application`).

> Is it possible to change the restart script to not pass the config-profile flag?

I could try that, and it's possible that would also fix the problem.  The reason I proposed the fix that I did is because I found the way the code is currently written to be a bit odd.  If the only reason we're waiting on the cluster setting to take effect is to schedule the backups, why do the waiting if we're not scheduling the backups.

**Comment by renatolabs** [comment]
>  If the only reason we're waiting on the cluster setting to take effect is to schedule the backups

While that is the only reason currently, one thing we will likely do in the near future (just as an example), is to create a default admin user on the tenant as well (currently, `createAdminUserForSecureClusters` only creates an admin on the system tenant.) At that point, we'll have to undo this change, we won't have a way to "skip" it on restarts.

> it's possible that would also fix the problem. The reason I proposed the fix

The way I understand it, this is not really fixing the issue, but skipping the problematic check. And the reason the check is problematic (instead of being a no-op as it is in "regular" clusters) is because we're using `config-profile=replication-source` while also changing the default cluster to be `system`. IMO, this is quite an unusual deployment model and not something we'd recommend doing in practice (correct me if I'm wrong) so I'm not sure roachprod should support it.

**Comment by ajstorm** [comment]
> The way I understand it, this is not really fixing the issue, but skipping the problematic check. And the reason the check is problematic (instead of being a no-op as it is in "regular" clusters) is because we're using config-profile=replication-source while also changing the default cluster to be system.  IMO, this is quite an unusual deployment model and not something we'd recommend doing in practice (correct me if I'm wrong) so I'm not sure roachprod should support it.

I agree that it doesn't fully resolve the issue (and I'm happy to close this PR as a result).  That being said, I wonder if this mode of starting up may be more realistic than we're anticipating.  The PCR docs say that nodes should be started with a `config-profile` set.  They don't say currently say anything about how the `config-profile` need only be applied at first node startup.  As a result, it's not unreasonable that users will apply a `config-profile` on upgrade as well.  At the same time, there's nothing preventing users for changing the `default_target_cluster` though I'd admit it's likely to be rare.

Is there a reason why when `config-profile=replication-source` that roachprod doesn't directly apply the necessary configuration to the `application` tenant as opposed to waiting for `default_target_cluster` to move away from `system`?  That would also fix the problem and would be more faithful to the desire to always have backup schedules (and any other operations) apply to the application tenant.  Is it because it would result in a more obtrusive change to the code?

**Comment by renatolabs** [comment]
> They don't say currently say anything about how the `config-profile` need only be applied at first node startup.  As a result, it's not unreasonable that users will apply a `config-profile` on upgrade as well.  At the same time, there's nothing preventing users for changing the `default_target_cluster` though I'd admit it's likely to be rare.

Note that _neither_ passing `config-profile` when restarting nor renaming the default tenant are the problem (I did both in my first message in this PR).

The problem is switching the default tenant back to `system`, which I think is the most unlikely: the user has chosen `replication-source` (indicating they want to use PCR), but have also changed the cluster's default tenant back to system.

> doesn't directly apply the necessary configuration to the `application` tenant as opposed to waiting for `default_target_cluster` to move away from `system`?

I'm not sure  I understand the question -- are you suggesting we just read the `default_target_cluster` setting, and apply config changes to that cluster?

If so, this is not possible because, when we first bootstrap a cluster with `config-profile=replication-source`, it takes a few seconds until `default_target_cluster` changes to point to the application tenant. If we just read that setting and applied changes to whatever it returns, it would be impossible to distinguish between 1) a delay in updating the cluster setting on bootstrap vs 2) a user genuinely changing the default cluster to `system`.

**Comment by ajstorm** [comment]
Closing based on offline conversation.  We can revisit this issue after we conclude any 24.1 work on the config profiles.
